clippy_lints/src/types/redundant_allocation.rs

```diff
@@ -54,8 +55,8 @@ pub(super) fn check(cx: &LateContext<'_>, hir_ty: &hir::Ty<'_>, qpath: &QPath<'_
     };
     let inner_span = match qpath_generic_tys(inner_qpath).next() {
         Some(ty) => {
-            // Box<Box<dyn T>> is smaller than Box<dyn T> because of wide pointers
-            if matches!(ty.kind, TyKind::TraitObject(..)) {
+            // Reallocation of a fat pointer causes it to become thin
+            if !hir_ty_to_ty(cx.tcx, ty).is_sized(cx.tcx.at(ty.span), cx.param_env) {
```

Comment:
Can you add a comment here that `hir_ty_to_ty` is never called inside a body, because the `check_ty` function in `mod.rs` is only called on types outside bodies and that this lint is not run on locals (also guaranteed in `mod.rs`).